torchmultimodal/models/gpt.py

```diff
@@ -28,12 +29,278 @@ class TransformerLayerOutput(NamedTuple):
+    """
+
+    def __init__(
+        self,
+        d_model: int,
+        num_tokens: int,
+        latent_shape: Tuple[int, ...],
+        in_tokenizer: nn.Module,
+        out_tokenizer: nn.Module,
+        mm_decoder: nn.Module,
+        in_projection: Optional[nn.Module] = None,
+        out_projection: Optional[nn.Module] = None,
+        norm_layer: Optional[Callable[..., nn.Module]] = None,
+    ) -> None:
+        super().__init__()
+        if not all(
+            [hasattr(in_tokenizer, attr_name) for attr_name in ["encode", "lookup"]]
+        ):
+            raise AttributeError(
+                "Input modality tokenizer must have methods 'encode' and 'lookup'."
+            )
+
+        if not all(
+            [
+                hasattr(out_tokenizer, attr_name)
+                for attr_name in ["encode", "lookup", "decode"]
+            ]
+        ):
+            raise AttributeError(
+                "Output modality tokenizer must have methods 'encode', 'lookup' and 'decode'."
+            )
+
+        self.latent_shape = latent_shape
+        self.in_tokenizer = in_tokenizer
+        self.out_tokenizer = out_tokenizer
+        self.mm_decoder = mm_decoder
+        self.in_projection = in_projection
+        self.out_projection = out_projection
+        if norm_layer is None:
+            norm_layer = partial(nn.LayerNorm, eps=1e-5)
+        self.norm = norm_layer(normalized_shape=d_model)
+        self.to_logit = nn.Linear(d_model, num_tokens, bias=False)
+        # This will give us equal probabilities after the soft max layer initially to avoid biasing
+        # towards any particular prediction category
+        self.to_logit.weight.data.copy_(torch.zeros(num_tokens, d_model))
+
+    def forward(
+        self,
+        in_tokens: Optional[Tensor] = None,
+        out_tokens: Optional[Tensor] = None,
+        in_pos_ids: Optional[Tensor] = None,
+        out_pos_ids: Optional[Tensor] = None,
+        attn_mask: Optional[Tensor] = None,
+        head_mask: Optional[Tensor] = None,
+        logits_mask: Optional[Tensor] = None,
+        use_cache: bool = False,
+        causal: bool = False,
+        right_shift: bool = False,
+        return_attn_weights: bool = False,
+        return_hidden_states: bool = False,
+    ) -> MultimodalGPTOutput:
+        decoder_output = self.fwd(
+            in_tokens=in_tokens,
+            out_tokens=out_tokens,
+            in_pos_ids=in_pos_ids,
+            out_pos_ids=out_pos_ids,
+            attn_mask=attn_mask,
+            head_mask=head_mask,
+            use_cache=use_cache,
+            causal=causal,
+            right_shift=right_shift,
+            return_attn_weights=return_attn_weights,
+            return_hidden_states=return_hidden_states,
+        )
+
+        hidden_states = decoder_output.last_hidden_states
+        logits = self.head(hidden_states, logits_mask)
+
+        return MultimodalGPTOutput(decoder_output, logits)
+
+    def fwd(
```

Comment:
Question about how we are splitting things up here. It feels unintuitive to have a separate `fwd` method that is doing most, but not all, of the usual forward. I was gonna suggest calling it decoder or something, but that is already an overloaded term here. Wondering if it makes sense to just add a util method that will return both in and out modality (where one can be None), then call `self.mm_decoder` in the main forward method. Then you don't have to pass all these args, and the core logic is clear in forward.